### PR TITLE
Feat/model

### PR DESCRIPTION
**Pull request by nenikitov**

### CODE REVIEW

File: engine/src/asset/model/dat/triangle.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    }
+}
+
+pub struct ModelTriangle {
+    pub points: [ModelPoint; 3],
+}
+
+impl ModelTriangle {
+    pub fn parse(texture_width: u32, texture_height: u32) -> impl Fn(&[u8]) -> Result<Self> {
```
Comment (UserIsntAvailable) [inline engine/src/asset/model/dat/triangle.rs:31]: nit: FnOnce

File: engine/src/asset/model/mod.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                    .iter()
+                    .map(|v| format!("    {v}"))
+                    .join(",\n"),
+                self.sequences
+                    .iter()
+                    .map(|s| format!("    {s}"))
+                    .join(",\n")
+            )
+        }
+    }
```
Comment (UserIsntAvailable) [inline engine/src/asset/model/mod.rs:236]: Why???? Just WHY???

**Review by UserIsntAvailable — CHANGES_REQUESTED** [review]
Everything was ok other than the mess `to_py` mess. Good Job :taco: :tada:.

File: engine/src/asset/model/dat/frame.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    pub vertices: Vec<ModelVertexParsed>,
+    pub triangle_normal_indexes: Vec<u8>,
+}
+
+impl ModelFrame {
+    pub fn parse(
+        vertex_count: usize,
+        triangle_count: usize,
+        frame_size: usize,
+    ) -> impl Fn(&[u8]) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (nenikitov) [reply engine/src/asset/model/dat/frame.rs:93]: Nom implements the trait `Parser` only for `FnMut`, so it's impossible to use `FnOnce` here

File: engine/src/utils/nom.rs [hunk 4]
```diff
@@ -12,10 +12,34 @@ macro_rules! re_export {
     };
 }
 
+pub mod number {
```
Comment (UserIsntAvailable) [inline engine/src/utils/nom.rs:15]:
```suggestion
/// Re-exports all `nom::number` items.
pub mod number {
```

File: engine/src/asset/model/mod.rs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        model::dat::{
+            frame::ModelFrame,
+            header::ModelHeader,
+            sequence::{ModelSequence, ModelSequenceParsed},
+        },
+        AssetChunk,
+    },
+    error,
+    utils::nom::*,
+};
```
Comment (UserIsntAvailable) [inline engine/src/asset/model/mod.rs:19]:
```suggestion
use super::{Asset, AssetChunk, Extension, Kind};
use crate::{error, utils::nom::*};
use dat::{
    frame::ModelFrame,
    header::ModelHeader,
    sequence::{ModelSequence, ModelSequenceParsed},
    triangle::ModelTriangle,
};
use itertools::Itertools;
```

**Review by UserIsntAvailable — CHANGES_REQUESTED** [review]
Looks good; only some minor stuff. The PR is still marked as `draft`; idk if you want to merge as is, but I wouldn't mind. We can take care of `to_py` later.

File: engine/src/asset/model/mod.rs (hunk 2 above)
Comment (nenikitov) [reply engine/src/asset/model/mod.rs:238]: I added TODOs how to refactor it, which can be done later in a separate PR.

**Review by UserIsntAvailable — APPROVED** [review]
LGTM

**Pull request merged**